src/python-samplerate.cpp

```diff
@@ -0,0 +1,570 @@
+
+ private:
+  std::string message = "";
+};
+
+int get_converter_type(const py::object &obj) {
+  if (py::isinstance<py::str>(obj)) {
+    py::str py_s = obj;
+    std::string s = static_cast<std::string>(py_s);
+    if (s.compare("sinc_best") == 0) {
+      return 0;
+    } else if (s.compare("sinc_medium") == 0) {
+      return 1;
+    } else if (s.compare("sinc_fastest") == 0) {
+      return 2;
+    } else if (s.compare("zero_order_hold") == 0) {
+      return 3;
+    } else if (s.compare("linear") == 0) {
+      return 4;
+    }
+  } else if (py::isinstance<py::int_>(obj)) {
+    py::int_ val = obj;
+    return static_cast<int>(val);
+  } else if (py::isinstance<ConverterType>(obj)) {
+    py::int_ c = obj.attr("value");
+    return static_cast<int>(c);
+  }
+
+  throw std::domain_error("Unsupported converter type");
+  return -1;
+}
+
+void error_handler(int errnum) {
+  std::cout << errnum << std::endl;
+  if (errnum > 0 && errnum < 24) {
+    throw ResamplingException(errnum);
+  } else if (errnum != 0) {  // the zero case is excluded as it is not an error
+    // this will throw a segmentation fault if we call src_strerror here
+    // also, these should never happen
+    throw std::runtime_error("libsamplerate raised an unknown error code");
+  }
+}
+
+class Resampler {
+ private:
+  SRC_STATE *_state = nullptr;
+  int _err_num = 0;
+
+ public:
+  int _converter_type = 0;
+  int _channels = 0;
+
+ public:
+  Resampler(const py::object &converter_type, int channels)
+      : _converter_type(get_converter_type(converter_type)),
+        _channels(channels) {
+    _state = src_new(_converter_type, _channels, &_err_num);
+  }
+
+  // copy constructor
+  Resampler(const Resampler &r)
+      : _converter_type(r._converter_type), _channels(r._channels) {
+    _state = src_clone(r._state, &_err_num);
+    error_handler(_err_num);
+  }
+
+  ~Resampler() { src_delete(_state); }
+
+  py::array_t<float, py::array::c_style> process(
+      py::array_t<float, py::array::c_style | py::array::forcecast> input,
+      double sr_ratio, bool end_of_input) {
+    // accessors for the arrays
+    py::buffer_info inbuf = input.request();
+
+    // set the number of channels
+    int channels = 1;
+    if (inbuf.ndim == 2)
+      channels = inbuf.shape[1];
+    else if (inbuf.ndim > 2)
+      throw std::domain_error("Input array should have at most 2 dimensions");
+
+    if (channels != _channels || channels == 0)
+      throw std::domain_error("Invalid number of channels in input data.");
+
+    const size_t new_size = size_t(std::ceil(inbuf.shape[0] * sr_ratio));
+
+    // allocate output array
+    std::vector<size_t> out_shape{new_size};
+    if (inbuf.ndim == 2) out_shape.push_back(size_t(channels));
+    auto output = py::array_t<float, py::array::c_style>(out_shape);
+    py::buffer_info outbuf = output.request();
+
+    // libsamplerate struct
+    SRC_DATA src_data = {
+        static_cast<float *>(inbuf.ptr),   // data_in
```

Comment:
I see, thanks for pointing this out.

I think this would be mostly for readability, so it may be nice to keep the requirement lower and include more users (even if few).

Reply:
Agreed, it would be just for readability.